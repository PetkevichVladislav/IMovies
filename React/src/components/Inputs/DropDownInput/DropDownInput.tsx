import "../Input.scss";
import "./DropDownInput.scss";

import { FC, ReactNode, useState } from "react";

export interface IDropDownInput {
    placeholder: string;
    children?: ReactNode;
}

export const DropDownInput: FC<IDropDownInput> = ({ placeholder, children }) => {
    const [isExpanded, setIsDropDownExpanded] = useState<boolean>(false);

    const onDropDownClicked = () => {
        setIsDropDownExpanded(!isExpanded);
    };

    return (
        <div className="drop-down-input">
            <div className="drop-down-input__select-container" onClick={onDropDownClicked} data-testid="drop-down-container">
                <select className="input drop-down-input__select">
                    <option>{placeholder}</option>
                </select>
                <span className={`drop-down-input__appearance icon-arrow-down ${isExpanded ? "rotated" : ""}`}></span>
                <div className="drop-down-input__content-container"></div>
            </div>
            {isExpanded &&
                <div className="drop-down-input__content" data-testid="drop-down-content">
                    {children}
                </div>
            }
        </div>
    )
}


