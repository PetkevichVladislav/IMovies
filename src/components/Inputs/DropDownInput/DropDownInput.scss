.drop-down-input{
    &__select-container{
        position: relative;
    }

    &__select {
        width: 100%;
        appearance: none;
    }

    &__appearance{
        position: absolute;
        right: 40px;
        top: 20px;
    
        &::before {
            content: "\e900";
            display: inline-block;
            width: 0;
            height: 0;
            color: #F65261;
            font-size: 16px;
        }

        &.rotated{
            right: 15px;
            transform: rotate(180deg);
        }
    }

    &__content-container{
        position: absolute;
        left: 0;
        right: 0;
        top: 0;
        bottom: 0;
    }

    &__content {
        display: block;
        outline: #F65261 2px solid;
        background-color: #232323;
        border: none;
        border-radius: 4px;
        padding: 18px 16px;

        label {
            display: block;
        }
    }
}

