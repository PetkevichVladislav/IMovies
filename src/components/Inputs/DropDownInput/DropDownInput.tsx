import { FC, ReactNode, useEffect, useMemo, useState } from "react";

import "./DropDownInput.scss";
import "../Input.scss";
import React from "react";
import { transform } from "typescript";

export interface IDropDownInput {
    placeholder: string;
    children?: ReactNode;
}

export const DropDownInput: FC<IDropDownInput> = ({ placeholder, children}) => {
    const [isExpanded, setIsDropDownExpanded] = useState(false);

    const onDropDownClicked = () => {
        setIsDropDownExpanded(!isExpanded);
    };

    // useEffect(() => {
    //     const handleClickOutside = (event: MouseEvent) => {
    //         ["drop-down-input__content", 'drop-down-input__content']
    //         const a = (event.target as Element).className === ;
    //         console.log((event.target as Element).className);
    //         if (isExpanded && !a) {
    //             setIsDropDownExpanded(false);
    //         }
    //     };

    //     document.addEventListener('click', (event) => handleClickOutside(event), true);
    //     return () => {
    //         document.removeEventListener('click', handleClickOutside, true);
    //     };
    // }, [isExpanded, setIsDropDownExpanded]);
    
    return (
        <>
            <div className="drop-down-input__select-container" onClick={onDropDownClicked}>
                <select className="input drop-down-input__select">
                    <option>{placeholder}</option>
                </select>
                <span className={`drop-down-input__appearance icon-arrow-down ${isExpanded ? "rotated": ""}`}></span>
                <div className="drop-down-input__content-container"></div>
            </div>
            {isExpanded &&
                <div className="drop-down-input__content">
                    {children}
                </div>
            }
        </>
    )
}


