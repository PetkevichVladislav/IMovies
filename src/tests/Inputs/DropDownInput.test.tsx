import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { DropDownInput, IDropDownInput } from '../../components/Inputs/DropDownInput/DropDownInput';



describe("Drop Down Input", () => {
    const dropDownInputParams : IDropDownInput = {
        placeholder: "Select value",
        children: <div data-testid="children"></div>
    };

    it("When data is valid component should renders correctly", () =>{
        const tree = renderer.create(<DropDownInput {...dropDownInputParams}/>)
        .toJSON();
        expect(tree).toMatchSnapshot();
    });
    
    it("when user clicks on container it should show dpor down", () =>{
        render(<DropDownInput {...dropDownInputParams}/>);
        const container = screen.getByTestId("drop-down-container");
        fireEvent.click(container);
        const input = screen.getByTestId("drop-down-content");
        expect(input).toContainElement(screen.getByTestId("children"));
    });
});