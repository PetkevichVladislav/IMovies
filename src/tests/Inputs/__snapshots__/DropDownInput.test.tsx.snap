// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drop Down Input When data is valid component should renders correctly 1`] = `
<div
  className="drop-down-input"
>
  <div
    className="drop-down-input__select-container"
    data-testid="drop-down-container"
    onClick={[Function]}
  >
    <select
      className="input drop-down-input__select"
    >
      <option>
        Select value
      </option>
    </select>
    <span
      className="drop-down-input__appearance icon-arrow-down "
    />
    <div
      className="drop-down-input__content-container"
    />
  </div>
</div>
`;
